File: ospray_modules/optix/geom/Instance.cpp
/* Copyright (c) 2018, EPFL/Blue Brain Project
 * All rights reserved. Do not distribute without permission.
 * Responsible Author: Jonas Karlsson <jonas.karlsson@epfl.ch>
 *
 * This file is part of https://github.com/BlueBrain/ospray-modules
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <ospray/SDK/common/Data.h>

#include "../Model.h"
#include "../render/Material.h"
#include "Instance.h"

#include <optix.h>
#include <optixu/optixpp_namespace.h>
#include <optixu/optixu_math_namespace.h>
#include <optixu/optixu_matrix_namespace.h>

namespace bbp
{
namespace optix
{
std::string Instance::toString() const
{
    return "ospray::Instance";
}

Instance::~Instance()
{
    if (_transform)
        _transform->destroy();
}

void Instance::finalize(Model* parentModel)
{
    if (_transform)
        _transform->destroy();

    ospray::vec3f vx = getParam3f("xfm.l.vx", ospray::vec3f(0.f, 0.f, 0.f));
    ospray::vec3f vy = getParam3f("xfm.l.vy", ospray::vec3f(0.f, 0.f, 0.f));
    ospray::vec3f vz = getParam3f("xfm.l.vz", ospray::vec3f(0.f, 0.f, 0.f));
    ospray::vec3f p = getParam3f("xfm.p", ospray::vec3f(0.f, 0.f, 0.f));

    auto subModel = (Model*)getParamObject("model", nullptr);

    const float matrixData[16] = {vx.x, vx.y, vx.z, p.x, //
                                  vy.x, vy.y, vy.z, p.y, //
                                  vz.x, vz.y, vz.z, p.z, //
                                  0.0f, 0.0f, 0.0f, 1.0f};
    _transform = subModel->instance(::optix::Matrix4x4(matrixData));

    parentModel->addTransformInstance(_transform);
}
}
}

File: python/brayns/instance/instance.py
# Copyright (c) 2015-2022 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from typing import Any, Optional, Protocol, TypeVar

from brayns.instance.jsonrpc.json_rpc_id import JsonRpcId
from brayns.instance.jsonrpc.json_rpc_request import JsonRpcRequest
from brayns.instance.request_future import RequestFuture

T = TypeVar('T', bound='Instance')


class Instance(Protocol):

    def __enter__(self: T) -> T:
        return self

    def __exit__(self, *_) -> None:
        self.disconnect()

    def disconnect(self) -> None:
        pass

    def request(self, method: str, params: Any = None) -> Any:
        task = self.task(method, params)
        return task.wait_for_result()

    def task(self, method: str, params: Any = None) -> RequestFuture:
        id = 0
        while self.is_running(id):
            id += 1
        request = JsonRpcRequest(id, method, params)
        return self.send(request)

    def is_running(self, id: JsonRpcId) -> bool:
        return False

    def send(self, request: JsonRpcRequest) -> RequestFuture:
        raise NotImplementedError()

    def poll(self, block: bool = True, timeout: Optional[float] = None) -> None:
        pass

    def cancel(self, id: JsonRpcId) -> None:
        pass
